fix(compiler): fix i18n IDs generation, add a migration tool

ocombe opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added


**What kind of change does this PR introduce?**
- [x] Bugfix


**What is the current behavior?**
See #15573.


**What is the new behavior?**
When we extract messages we digest them to generate ids (unless the id is provided).
With this fix we ignore the content of the placeholder when we digest to only use the fact that it's a placeholder and its position.
We already did the same thing with ICU expressions for XMB and this previous fix is now also applied to ICU expressions for XLIFF (now that XLIFF supports ICU expressions).
Example previously `<p i18n>{{expr}}</p>`, `<p i18n>{{ expr }}</p>` and `<p i18n>{{expr2}}</p>` had different ids but the exact same content once extracted. With this fix those 3 messages will only be extracted once with the same id.
We also aligned the digest functions for all serializers, they will all use the decimal digest (instead of sha1 previously for xliff).

-----

Because of the bug #15573 we had to make a breaking change in the way i18n ids are generated.
To mitigate the problem, a new flag can be used to set the version of the digest method used: `--i18nVersion` with values 0 (old, default for v4) or 1 (new, default for v5) can be used by the cli tools for extraction (ng-xi18n) and AOT (ngc). The token `I18N_VERSION` can be used in your code for JIT.
We also added a new migration cli tool in @angular/compiler-cli named ng-migrate-i18n.
Usage is the following: `ng-migrate-i18n --files src/i18n/*.xlf --i18nFormat xlf`
--files: (required) the translation files that you want to migrate, you can use glob expressions to migrate multiple files at the same time (as long as they use the same format)
--i18nFormat: (required) the format of the translation files that you want to migrate
--i18nVersion: (optional) the version from which you are migrating. The only value for now is "0" (the new version is "1")
--mapping: (optional) if you add this option it will generate a json file of mappings {newId --> [oldIds]} instead of migrating the translation files
--resolve: (optional) "manual" by default, you can use "auto" so that the cli tool doesn't ask when it finds conflicts (in which case it will take the first value)


**Does this PR introduce a breaking change?**
- [x] Yes

Unfortunately this is a breaking change because the auto-generated ids for extracted messages will change if the message contains placeholders (or ICU expressions for XLIFF)...
To mitigate the problem, a new flag can be used to set the version of the digest method used. It iss `--idVersion` with values 0 (old, default for v4) or 1 (new, default for v5) for extraction (ng-xi18n) and AOT (ngc), and the token `ID_VERSION` for JIT.
Fow now we still use the old method to generate the ids, but once v5 is released we will reverse the flag to only use the new version.

vicb commented on packages/compiler/test/i18n/digest_spec.ts:
```diff
@@ -21,6 +24,27 @@ export function main(): void {
+        const sourceId = digest(extractFirstMsg(HTML1));
+
+        expect(sourceId).toEqual('c11e34173f455c06a2ecd5859038917ca52780b1');
+        expect(digest(extractFirstMsg(HTML2))).toEqual(sourceId);
```
Not sure if we should `expect(sourceId)` here

ocombe replied:
not really, I can remove it

vicb commented on packages/compiler/test/i18n/extractor_merger_spec.ts:
```diff
@@ -77,9 +77,7 @@ export function main() {
                 '<div i18n="m|d">{count, plural, =0 { <p i18n-title i18n-desc title="title" desc="desc"></p>}}</div>'))
             .toEqual([
               [
-                [
```
Copy the former serializer to tests and use it. The expectations should not be modified here.

ocombe replied:
ahah damn, that's what I did the first time and then I thought that we should remove it because it was only used in the tests

vicb commented on packages/compiler/test/i18n/i18n_parser_spec.ts:
```diff
@@ -94,7 +94,7 @@ export function main() {
               [['msg'], 'm', 'd'],
               [
                 [
-                  '{count, plural, =0 {[<ph tag name="START_PARAGRAPH"><ph tag name="START_BOLD_TEXT"></ph name="CLOSE_BOLD_TEXT"></ph name="CLOSE_PARAGRAPH">]}}'
```
same as previous file

vicb commented on packages/compiler/test/i18n/serializers/i18n_ast_spec.ts:
```diff
@@ -19,8 +19,7 @@ export function main(): void {
         const text = serializeNodes(nodes).join('');
-        expect(text).toEqual(
-            'b<ph icu name="ICU">{count, plural, =0 {[{sex, select, male {[m]}}]}}</ph>a');
+        expect(text).toEqual('b<ph icu name="ICU">{plural, =0 {[{select, male {[m]}}]}}</ph>a');
```
same comment

vicb commented on packages/compiler/test/i18n/serializers/xtb_spec.ts:
```diff
@@ -83,7 +83,7 @@ export function main(): void {
           '7717087045075616176': `*<ph name="ICU"/>*`,
           '5115002811911870583':
-              `{VAR_PLURAL, plural, =1 {[<ph name="START_PARAGRAPH"/>, rab, <ph name="CLOSE_PARAGRAPH"/>]}}`,
+              `{plural, =1 {[<ph name="START_PARAGRAPH"/>, rab, <ph name="CLOSE_PARAGRAPH"/>]}}`,
```
same comment

vicb commented:
@ocombe thanks for the PR - I have added a few comments.

We need to come with a migration tool before merging this PR, can you work on it ?

vicb reviewed (dismissed):
see inline comments

ocombe commented:
I changed the code that you requested @vicb, let me know if it's ok now

vicb commented on packages/compiler/src/i18n/serializers/serializer.ts:
```diff
@@ -9,6 +9,8 @@
 import * as i18n from '../i18n_ast';
 
 export abstract class Serializer {
+  constructor(public idVersion?: string|null) {}
```
why `public idVersion?: string|null` on an abstract class ?

vicb commented on packages/compiler/src/i18n/serializers/serializer.ts:
```diff
@@ -17,7 +19,7 @@ export abstract class Serializer {
       {locale: string | null, i18nNodesByMsgId: {[msgId: string]: i18n.Node[]}};
 
-  abstract digest(message: i18n.Message): string;
+  abstract digest(message: i18n.Message, idVersion?: string|null): string;
```
do we need the version here if serializer have a version ?

ocombe replied:
In the migration process we need to generate old ids and new ids with the same serializer. But we can also create 2 serializers with a different id version.
I was thinking that maybe we needed this for tests too, but I just checked and we use the exported digest version directly.
I'll remove this.

vicb commented on packages/compiler/src/i18n/serializers/xliff2.ts:
```diff
@@ -100,7 +107,12 @@ export class Xliff2 extends Serializer {
+    if (!idVersion || idVersion === '0') {
+      return decimalDigest(message);
+    }
+    return digest(message, idVersion || this.idVersion);
```
seems like the serializer and the digest IDs are decoupled ?
so does it make sense to use the same version number ?

ocombe replied:
Yes, there's a potential bug here, I'll fix it

vicb commented on packages/compiler/src/i18n/serializers/xmb.ts:
```diff
@@ -85,8 +92,12 @@ export class Xmb extends Serializer {
+    if (!idVersion || idVersion === '0') {
+      return decimalDigest(message);
+    }
+    return digest(message, idVersion || this.idVersion);
```
same comment as for xliff2

vicb commented on packages/compiler/test/i18n/extractor_merger_spec.ts:
```diff
@@ -477,9 +514,9 @@ function parseHtml(html: string): html.Node[] {
-function fakeTranslate(
-    content: string, implicitTags: string[] = [],
-    implicitAttrs: {[k: string]: string[]} = {}): string {
+function fakeTranslate(content: string, idVersion = '0'): string {
```
missing type for idVersion

ocombe replied:
Typescript doesn't need a type if there's a default typed value, it will infer it

vicb commented on packages/compiler-cli/integrationtest/src/messages_migrated.fr.xlf:
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
use the version number in file name ?

ocombe replied:
ah sorry you were talking about the name of the file, not the version on the xml element :D

vicb commented on packages/compiler-cli/src/extractor.ts:
```diff
@@ -48,58 +45,24 @@ export class Extractor {
-
-    switch (format) {
-      case 'xmb':
-        serializer = new compiler.Xmb();
```
shouldn't those have a version from the previous commit ?

ocombe replied:
It was using the default value before

vicb commented on packages/compiler-cli/src/extractor.ts:
```diff
@@ -27,17 +27,14 @@ export class Extractor {
+    return this.extractBundle().then((bundle: compiler.MessageBundle) => {
+      const serializer = compiler.createSerializer(formatName);
+      const content = this.serialize(bundle, serializer);
+      const dstFile = outFile || `messages.${serializer.extension}`;
```
independent change -> independent commit

vicb commented on packages/compiler-cli/src/migrator.ts:
```diff
@@ -0,0 +1,55 @@
+import {Extractor} from './extractor';
+import {PathMappedCompilerHost} from './path_mapped_compiler_host';
+
+export class Migrator {
```
add comment on what the class is supposed to do

vicb commented on packages/compiler-cli/src/migrator.ts:
```diff
@@ -0,0 +1,55 @@
+      private options: tsc.AngularCompilerOptions, public host: ts.CompilerHost,
+      private extractor: Extractor, private i18nFile: string, private outFile?: string|null) {}
+
+  migrate(formatName: string|null, idVersion: string|null): Promise<void|string> {
```
add comment ?
is version from or to ? does it make sense for it to be null ?

ocombe replied:
idVersion is always "from", because the version we're migrating to is always the latest, I'll add comments.
It can be null because cli parameters are null when not provided.

vicb commented on packages/compiler-cli/src/ngtools_api.ts:
```diff
@@ -20,52 +20,51 @@ import * as ts from 'typescript';
 import {CodeGenerator} from './codegen';
```
Have changes in this file been validtaed with @hansl ?

ocombe replied:
It's only a refactoring to mutualize common parameters, but all existing interfaces still work as they did before, there's no change necessary in the cli for that.
The only difference maybe is that `extractI18n` returns a `Promise<string>` instead of `void`, because it allows us to have better tests, but I can make that change just for the migration part if necessary

vicb commented on packages/compiler/src/i18n/message_bundle.ts:
```diff
@@ -23,8 +23,7 @@ export class MessageBundle {
 
   constructor(
       private _htmlParser: HtmlParser, private _implicitTags: string[],
-      private _implicitAttrs: {[k: string]: string[]}, private _locale: string|null = null,
```
revert from the previous commit ?

ocombe replied:
Yes, I should have amended the previous commit instead. I've moved the idVersion property to the serializer so that we don't need to pass the option around (and change the method signatures)

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,114 @@
+
+
+/**
+ * Migrate i18n messages from one id version to the latest
```
is it "the latest" or a fixed version ?

ocombe replied:
Always the latest, there's no option to migrate to a specific version (the cli parameter `idVersion` refers to the "from", not the "to").
Do you want an option to specify the "to" ? I don't think that it's necessary

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+}
+
+class _MultiFormatVisitor implements ml.Visitor {
```
remove leading `_`

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,114 @@
+        break;
+
+      default:
+        // Recurse on unhandled nodes
```
"unhandled" ? 
I think you can remove the comment altogether

vicb commented on packages/compiler/src/i18n/serializers/xmb.ts:
```diff
@@ -15,7 +15,7 @@ import * as xml from './xml_helper';
 const _MESSAGE_TAG = 'msg';
 const _PLACEHOLDER_TAG = 'ph';
-const _EXEMPLE_TAG = 'ex';
+const _EXAMPLE_TAG = 'ex';
```
nice catch but this should be part of an other commit

ocombe replied:
ok :)

vicb reviewed (commented):
See inline comments

vicb commented on packages/compiler-cli/package.json:
```diff
@@ -6,7 +6,8 @@
     "ngc": "./src/main.js",
-    "ng-xi18n": "./src/extract_i18n.js"
+    "ng-xi18n": "./src/extract_i18n.js",
+    "ng-mi18n": "./src/migrate_i18n.js"
```
Not sure if a short name is valuable for this ?

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,114 @@
+  migrate(translations: string, formatName?: string|null, idVersion?: string|null): string {
+    const serializer = createSerializer(formatName);
+    const oldIdsToNewIds = new Map<string, string>();
+    const newIdsToOldIds = new Map<string, string>();
```
a new ID can map to several old IDs, no ?

ocombe replied:
That's why I have 2 maps, the first one is "all olds ids --> new id", the second one is used for deduplication, so there's only one "new id / old id" couple. If we had 2 old translations, then we only take the first one and ignore all the others.

It was easier to do it like this than to have only one map "new id / [old ids]" where we would have had to loop over all ids to find if it was in the list.

ocombe commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -10,18 +10,26 @@ import {utf8Encode} from '../util';
 
-export function digest(message: i18n.Message): string {
-  return message.id || sha1(serializeNodes(message.nodes).join('') + `[${message.meaning}]`);
+export function digest(message: i18n.Message, idVersion?: string | null): string {
```
We can receive the value from the cli where it can be null, otherwise we have to fix the default version at the different entry points (I prefer to only have it in one place).

ocombe commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -10,18 +10,26 @@ import {utf8Encode} from '../util';
+  }
+
+  switch (idVersion) {
+    case '1':
```
Good point, I'll do that. And no it's not documented yet, I'll add it too.

ocombe commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -80,6 +88,30 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
 
+const serializerVisitorIgnoreIcuExp = new _SerializerIgnoreIcuExpVisitor();
+
+export function serializeNodesIgnoreIcuExp(nodes: i18n.Node[]): string[] {
```
Hmm I don't see a reason now. Probably something that I changed along the way and reversed. I'll remove this change since it isn't necessary anymore.

ocombe commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -80,6 +88,30 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
+ *
+ * @internal
+ */
+class _SerializerV2 extends _SerializerIgnoreIcuExpVisitor {
```
True, at first I had versions 0, 1 and 2, and then refactored the code to only have version 0 and 1, I forgot to change the name of the serializer, I'll do that.
Ok for the rename of "private" classes too, I'll do that later.

ocombe commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -80,6 +88,30 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
+
+const serializerVisitorV2 = new _SerializerV2();
+
+export function serializeNodesV2(nodes: i18n.Node[]): string[] {
```
Agreed, we don't need it.

ocombe commented on packages/compiler/test/i18n/extractor_merger_spec.ts:
```diff
@@ -421,29 +437,49 @@ export function main() {
         const HTML = `<p i18n-title="m|d" title="foo"></p>`;
         expect(fakeTranslate(HTML)).toEqual('<p title="**foo**"></p>');
+        expect(fakeTranslate(HTML, '1')).toEqual('<p title="**foo**"></p>');
+        expect(fakeTranslate(HTML, '2')).toEqual('<p title="**foo**"></p>');
```
There is no version 2 anymore, I need to remove this

mary-poppins commented:
The angular.io preview for a710c3cba1096028f565d2651e49260449dc9718 is available [here][1].

[1]: https://pr15621-a710c3cba1096028f565d2651e49260449dc9718.ngbuilds.io/

ocombe committed:
refactor(compiler): change xliff serializer digest function to decimal

BREAKING CHANGE: updates the xliff serializer to use the same digest function. This is a breaking change because the IDs generated won't match with your existing translations. We provide a migration tool to update your existing files to the new method.

ocombe committed:
fix(compiler): ignore ph/ICU expressions when generating i18n ids

BREAKING CHANGE: This fix updates the digest function to ignore the content of placeholders.
Example previously `<p i18n>{{expr}}</p>`, `<p i18n>{{ expr }}</p>` and `<p i18n>{{expr2}}</p>` had different ids but the exact same content once extracted. With this fix those 3 messages will only be extracted once with the same id.
This is a breaking change because the auto-generated ids for extracted messages will change if the message contains placeholders... To mitigate the problem, a migration tool will be provided.

Fixes #15573

mary-poppins commented:
The angular.io preview for 7ea549e8d78ec86cd8d01e778b25b161ed1a03d8 is available [here][1].

[1]: https://pr15621-7ea549e8d78ec86cd8d01e778b25b161ed1a03d8.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 465b5f9a5a2c0ec94cebb797bfe6bf9bf7383a9b is available [here][1].

[1]: https://pr15621-465b5f9a5a2c0ec94cebb797bfe6bf9bf7383a9b.ngbuilds.io/

mary-poppins commented:
The angular.io preview for e390595a6ec2031ccb5abebd86a463fd7b74289f is available [here][1].

[1]: https://pr15621-e390595a6ec2031ccb5abebd86a463fd7b74289f.ngbuilds.io/

mary-poppins commented:
The angular.io preview for a82ca97b4758591550919d3a45dcecd9e7b60b16 is available [here][1].

[1]: https://pr15621-a82ca97b4758591550919d3a45dcecd9e7b60b16.ngbuilds.io/

vicb commented on packages/compiler/src/i18n/serializers/xliff.ts:
```diff
@@ -8,7 +8,7 @@
 
 import * as ml from '../../ml_parser/ast';
```
Commit comment:
- is this a feat or a refactor ?
- remove: "all i18n serializer are using the decimal digest function to generate the IDs. This PR "

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -81,6 +81,24 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
+ *
+ * Ignore the PH/ICU expressions so that message IDs stays identical if only the expression changes.
+ *
+ * @internal
```
I don't think you need `@internal` because the class is not export.

I would also put the comment inside the method to explain the diff with the base visitor.

Is "Uses Id generation version 1" usefull ?

ocombe replied:
I've used `@internal` because it was on all the other similar methods already

vicb replied:
The fix is the other way around: remove from all non exported class

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -81,6 +81,24 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
+
+const serializerVisitorV1 = new SerializerV1();
+
+function serializeNodesV1(nodes: i18n.Node[]): string[] {
```
could you inline this ? (ie same thing you revert after my previous round of comments)

vicb commented on packages/compiler/src/i18n/serializers/xliff.ts:
```diff
@@ -112,7 +112,7 @@ export class Xliff extends Serializer {
   }
 
-  digest(message: i18n.Message): string { return decimalDigest(message); }
+  digest(message: i18n.Message): string { return digest(message, this.idVersion); }
```
Once again: I don't see why digest and idVersion are coupled.

digest as 2 algos and serializers have 2 versions (and more later). 
digest should not be aware about idVersion

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -81,6 +81,24 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
+ *
+ * @internal
+ */
+class SerializerV1 extends _SerializerIgnoreIcuExpVisitor {
```
& also same comment as for the digest: the version is for the serializers remove the V1 for something meaningful in the digest context. (cf "IgnoreIcuExp")

vicb commented on packages/compiler/src/i18n/serializers/xliff2.ts:
```diff
@@ -100,7 +100,9 @@ export class Xliff2 extends Serializer {
 
-  digest(message: i18n.Message): string { return decimalDigest(message); }
+  digest(message: i18n.Message): string {
+    return digest(message, this.idVersion);
```
same comment, digest should not care about the version

vicb commented on packages/compiler/test/i18n/digest_spec.ts:
```diff
@@ -22,6 +22,21 @@ export function main(): void {
+        const HTML3 =
+            `<div i18n>some element {{placeholder2}} and {count2, plural, =0 {zero} =1 {one} =2 {two} other {<b>many</b>}}</div>`;
+
+        const sourceId = digest(extractFirstMsg(HTML1));
```
const id = ...

mary-poppins commented:
The angular.io preview for ca7487da0074caa18d4b35e46de2a26729eede7a is available [here][1].

[1]: https://pr15621-ca7487da0074caa18d4b35e46de2a26729eede7a.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 4beb50836240f1ec58d88753148ce252d65143dd is available [here][1].

[1]: https://pr15621-4beb50836240f1ec58d88753148ce252d65143dd.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 1a260ed76c10d69ff6255d41f60fe55a8c6d9204 is available [here][1].

[1]: https://pr15621-1a260ed76c10d69ff6255d41f60fe55a8c6d9204.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 36b46345d5a35a6fba503b538d8e58e7725d8499 is available [here][1].

[1]: https://pr15621-36b46345d5a35a6fba503b538d8e58e7725d8499.ngbuilds.io/

mary-poppins commented:
The angular.io preview for a983c2c15fac5160f22b17d6039078b025eac745 is available [here][1].

[1]: https://pr15621-a983c2c15fac5160f22b17d6039078b025eac745.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 87a214b57ec766d78a72a12b345e7b670d2e85d1 is available [here][1].

[1]: https://pr15621-87a214b57ec766d78a72a12b345e7b670d2e85d1.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5b8c7fa8711eaef50f24e7e5c9e4c1c553c695d8 is available [here][1].

[1]: https://pr15621-5b8c7fa8711eaef50f24e7e5c9e4c1c553c695d8.ngbuilds.io/

vicb commented on packages/compiler-cli/src/extract_i18n.ts:
```diff
@@ -8,22 +8,31 @@
+ * @param cliOptions
+ * @param program
+ * @param host
+ * @returns {Promise<void|string>}
```
remove useless annotation here

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -10,26 +10,60 @@ import {utf8Encode} from '../util';
+ * @returns {string}
+ *
+ * @internal
+ * @deprecated
```
add a comment about why& when this is deprecated

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -10,26 +10,60 @@ import {utf8Encode} from '../util';
+/**
+ * Digest with a sha1 function
+ * @param message
+ * @returns {string}
```
remove useless annotation everywhere

mary-poppins commented:
The angular.io preview for 25f12bca4dbe31115261eba1a1844b5d44a88132 is available [here][1].

[1]: https://pr15621-25f12bca4dbe31115261eba1a1844b5d44a88132.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 0edf117a72cc5d660cf27f3d458719473f40f863 is available [here][1].

[1]: https://pr15621-0edf117a72cc5d660cf27f3d458719473f40f863.ngbuilds.io/

vicb commented:
Commit comments:
- first commit: "updates the xliff serializer to use the same digest function" ?
- second commit: "This fix updates how the digest function to ignore the content of placeholders" ?
- third commit: "the translation file" (appear multiple times) -> it could be either a translation or a message file

What about removing the `--i18nVersion` parameter and renaming the script `ng-migrate-i18n-4to5` ?

vicb commented on packages/compiler/test/i18n/integration_xliff_spec.ts:
```diff
@@ -326,7 +326,7 @@ const XLIFF_EXTRACTED = `
       </trans-unit>
       <trans-unit id="i18n17" datatype="html">
-        <source>{VAR_PLURAL, plural, =0 {zero} =1 {one} =2 {two} other {<x id="START_BOLD_TEXT" ctype="x-b"/>many<x id="CLOSE_BOLD_TEXT" ctype="x-b"/>} }</source>
+        <source>\{VAR_PLURAL, plural, =0 \{zero} =1 \{one} =2 \{two} other \{<x id="START_BOLD_TEXT" ctype="x-b"/>many<x id="CLOSE_BOLD_TEXT" ctype="x-b"/>} }</source>
```
why do we have this change ?

ocombe replied:
we shouldn't

vicb replied:
Could you update the commit to remove them ?

vicb commented on packages/compiler-cli/integrationtest/test/test_ngtools_api.ts:
```diff
@@ -185,6 +190,161 @@ function i18nTest() {
       });
 }
 
+function i18nMigrationTest() {
```
add quick comment on what the function is testing

vicb commented on packages/compiler-cli/integrationtest/test/test_ngtools_api.ts:
```diff
@@ -185,6 +190,161 @@ function i18nTest() {
+      });
+}
+
+function i18nMigrationMappingTest(testName?: 'outFile' | 'i18nFile') {
```
add comment - what is outFile vs i18nFile ?

vicb replied:
also could you factor common test code ?

vicb commented on packages/compiler-cli/src/ngtools_api.ts:
```diff
@@ -146,10 +149,27 @@ export class NgTools_InternalApi_NG_2 {
+
+  /**
+   * @internal
+   * @private
```
remove `@private`

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -10,18 +10,36 @@ import {utf8Encode} from '../util';
+ * Used by Id generation version 0 for xlf
+ *
+ * @internal
+ * @deprecated since PR #15621, use decimalIgnorePhDigest instead
```
Should be `@deprecated from v5` (help user know when it is deprecated and core team when to remove the code)

vicb commented on packages/compiler/src/i18n/i18n_html_parser.ts:
```diff
@@ -25,11 +25,11 @@ export class I18NHtmlParser implements HtmlParser {
+      private _htmlParser: HtmlParser, translations?: string, translationsFormat?: string|null,
       missingTranslation: MissingTranslationStrategy = MissingTranslationStrategy.Warning,
-      console?: Console) {
+      console?: Console, version?: I18nVersion) {
```
make it required

vicb commented on packages/compiler/src/i18n/i18n_html_parser.ts:
```diff
@@ -55,20 +55,20 @@ export class I18NHtmlParser implements HtmlParser {
 }
 
-function createSerializer(format?: string): Serializer {
+export function createSerializer(format?: string | null, version?: I18nVersion): Serializer {
```
same here

vicb commented on packages/compiler/src/i18n/serializers/serializer.ts:
```diff
@@ -6,9 +6,15 @@
 import * as i18n from '../i18n_ast';
 
 export abstract class Serializer {
+  abstract version: I18nVersion;
```
should be consistent with the ctor arg IMO

vicb replied:
not sure if the version is needed on the base class

vicb commented on packages/compiler-cli/src/codegen.ts:
```diff
@@ -63,37 +63,59 @@ export class CodeGenerator {
+
+
+/**
+ * @internal
```
Should be consistent with the previous fn ?

vicb commented on packages/compiler-cli/src/extract_i18n.ts:
```diff
@@ -8,22 +8,26 @@
+  return Extractor
+      .create(
+          ngOptions, program, host, cliOptions.locale, undefined, undefined,
+          getI18nVersion(cliOptions.i18nVersion))
```
thought: we have `getMissingTranslationStrategy, getI18nVersion` that normalize CLI params, what if we do the same for the format and use it below - and rename the functions `normalizeXxx` ?

ocombe replied:
I renamed those functions, but refactoring other things should be in another commit/PR 🙄

vicb commented on packages/compiler-cli/src/extractor.ts:
```diff
@@ -7,34 +7,35 @@
       private program: ts.Program) {}
 
-  extract(formatName: string, outFile: string|null): Promise<string[]> {
+  extract(formatName: string|null, outFile: string|null): Promise<string[]> {
```
my previous comment... and we can remove `|null` from here

vicb commented on packages/compiler-cli/src/migrator.ts:
```diff
@@ -0,0 +1,74 @@
+      const migrator = new compiler.Migrator(bundle);
+      if (i18nMapping) {
+        const mapping = JSON.stringify(migrator.getMapping(formatName, version));
+        let dstPath = 'messages_mapping.json';
```
maybe it should be `i18nMapping=path/to/map.json` to make things simpler ?

vicb commented on packages/compiler/src/i18n/i18n_html_parser.ts:
```diff
@@ -25,11 +25,11 @@ export class I18NHtmlParser implements HtmlParser {
 
   constructor(
-      private _htmlParser: HtmlParser, translations?: string, translationsFormat?: string,
+      private _htmlParser: HtmlParser, translations?: string, translationsFormat?: string|null,
```
why adding `|null` here ?

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,140 @@
+ *
+ * @experimental
+ */
+export class Migrator {
```
I would make it less generic (ie `Migrator4to5`)

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,140 @@
+  }
+}
+
+class MultiFormatVisitor implements ml.Visitor {
```
`IdAttributeUpdateVistor` ?
Add a comment explaining what it does (ie update the value of the id attr in an XML file on a, b, c tags)

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,140 @@
+      throw new Error(`Translations parse errors:\n${errors.join('\n')}`);
+    }
+
+    msgIdToElement.reverse().forEach(([id, element]) => {
```
`reverse()` is useless here.

What needs to happen here is a proper node replacement, ie replacing AST nodes instead doing a string replace.

```
node1
node2
node3
```

if you first change node1 then the SourceSpan for node2 and node3 become invalid. Which is why when you replace AST nodes you should start by the end.

Makes sense ?

ocombe replied:
We will loose special attributes if I do that. A lot of translation softwares add their own custom attributes to keep track of translations

vicb replied:
You should replace only the value for the id attribute

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,140 @@
+  private _errors: I18nError[];
+  private _msgIdToElement: [string, ml.Element][];
+
+  parse(content: string, url: string):
```
parse -> generateMapping ?

vicb commented on packages/compiler/src/i18n/serializers/xliff.ts:
```diff
@@ -31,7 +34,14 @@ const _CONTEXT_TAG = 'context';
-  write(messages: i18n.Message[], locale: string|null): string {
+  version: I18nVersion;
+
+  constructor(version?: I18nVersion) {
```
make version mandatory

vicb commented on packages/compiler/src/i18n/serializers/xliff.ts:
```diff
@@ -109,10 +119,15 @@ export class Xliff extends Serializer {
 
-  digest(message: i18n.Message): string { return decimalIgnorePhDigest(message); }
+  digest(message: i18n.Message): string {
+    if (!this.version) {
```
why `!version` ? should be `I18nVersion.Version0` ?

ocombe replied:
true (even if in this case I18nVersion.Version0's value is 0)

vicb commented on packages/compiler/test/i18n/message_bundle_spec.ts:
```diff
@@ -42,6 +44,13 @@ export function main(): void {
 }
 
 class _TestSerializer extends Serializer {
+  public version: I18nVersion;
```
remove `public` (consistency)

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -26,6 +26,23 @@ export const TRANSLATIONS_FORMAT = new InjectionToken<string>('TranslationsForma
+ * @experimental i18n support is experimental.
+ */
+export enum I18nVersion {
+  // initial version
```
add "up to angular v4"

vicb commented on packages/core/src/i18n/tokens.ts:
```diff
@@ -26,6 +26,23 @@ export const TRANSLATIONS_FORMAT = new InjectionToken<string>('TranslationsForma
+export enum I18nVersion {
+  // initial version
+  Version0 = 0,
+  // version after first breaking change:
```
add "from v5.0"

vicb commented on tools/@angular/tsc-wrapped/src/cli_options.ts:
```diff
@@ -11,40 +11,70 @@ export class CliOptions {
+  public i18nFile: string|null;
+  public i18nVersion: string|null;
+  public outFile: string|null;
+  public i18nMapping: string|null;
```
remove `public`s

vicb commented:
The second commit has unresolved comments, could you please fix them ?

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,157 @@
+ */
+export function applyIdsMapping(
+    translations: string, mapping: {[oldId: string]: string | null}): string {
+  const msgIdToElement = this.getElementsMapping(translations);
```
`this.` should be removed ?

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,157 @@
+    const newId = mapping[id];
+    if (typeof newId !== 'undefined' && newId !== id) {
+      if (newId) {
+        translations = translations.replace(id, newId);
```
replace using `SourceSpan`

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,157 @@
+            translations.substr(0, element.startSourceSpan !.start.offset).lastIndexOf('>') + 1;
+        const elementContent =
+            translations.substring(previousElement, element.endSourceSpan !.end.offset);
+        translations = translations.replace(elementContent, '');
```
work with `SourceSpan` indexes exclusively (ie no `replace`)

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,157 @@
+export function getIdsMapping(
+    messageBundle: MessageBundle, formatName?: string, fromVersion?: I18nVersion):
+    {newToOld: {[newId: string]: string[]}, oldToNew: {[oldId: string]: string}} {
+  const oldSerializer = createSerializer(formatName, fromVersion || I18nVersion.Version0);
```
what other than `I18nVersion.Version0` could this be ?

ocombe replied:
For now, nothing

vicb commented on packages/compiler/src/i18n/serializers/serializer.ts:
```diff
@@ -9,6 +9,8 @@
 import * as i18n from '../i18n_ast';
 
 export abstract class Serializer {
+  constructor() {}
```
remove ?

vicb commented on packages/compiler/src/i18n/serializers/xliff.ts:
```diff
@@ -31,7 +34,9 @@ const _CONTEXT_TAG = 'context';
-  write(messages: i18n.Message[], locale: string|null): string {
+  constructor(public version: I18nVersion) { super(); }
+
+  write(messages: i18n.Message[], locale: string|null, targetLocale?: string|null): string {
```
what is `targetLocale` here ?

ocombe replied:
good question... removing this

vicb commented on packages/compiler/test/i18n/serializers/xliff_spec.ts:
```diff
@@ -99,22 +100,22 @@ const WRITE_XLIFF = `<?xml version="1.0" encoding="UTF-8" ?>
       </trans-unit>
       <trans-unit id="baz" datatype="html">
-        <source>{VAR_PLURAL, plural, =0 {{VAR_SELECT, select, other {<x id="START_PARAGRAPH" ctype="x-p"/>deeply nested<x id="CLOSE_PARAGRAPH" ctype="x-p"/>} } } }</source>
+        <source>\{VAR_PLURAL, plural, =0 \{\{VAR_SELECT, select, other \{<x id="START_PARAGRAPH" ctype="x-p"/>deeply nested<x id="CLOSE_PARAGRAPH" ctype="x-p"/>} } } }</source>
```
revert

mary-poppins commented:
The angular.io preview for 76deb4afd00f68f83a327eaeb8bdaf6489efe71e is available [here][1].

[1]: https://pr15621-76deb4afd00f68f83a327eaeb8bdaf6489efe71e.ngbuilds.io/

mary-poppins commented:
The angular.io preview for cd6d937906d431d8318a90484935c29a87968854 is available [here][1].

[1]: https://pr15621-cd6d937906d431d8318a90484935c29a87968854.ngbuilds.io/

ocombe committed:
feat(compiler-cli): add tool ng-migrate-i18n & i18nVersion param

Because of the bug #15573 we had to make a breaking change in the way i18n ids are generated.
To mitigate the problem, a new flag can be used to set the version of the digest method used: `--i18nVersion` with values 0 (old, default for v4) or 1 (new, default for v5) can be used by the cli tools for extraction (ng-xi18n) and AOT (ngc). The token `I18N_VERSION` can be used in your code for JIT.
We also added a new migration cli tool in @angular/compiler-cli named ng-migrate-i18n.
Usage is the following: `ng-migrate-i18n --files src/i18n/*.xlf --i18nFormat xlf`
--files: (required) the translation files that you want to migrate, you can use glob expressions to migrate multiple files at the same time (as long as they use the same format)
--i18nFormat: (required) the format of the translation files that you want to migrate
--i18nVersion: (optional) the version from which you are migrating. The only value for now is "0" (the new version is "1")
--mapping: (optional) if you add this option it will generate a json file of mappings {newId --> [oldIds]} instead of migrating the translation files
--resolve: (optional) "manual" by default, you can use "auto" so that the cli tool doesn't ask when it finds conflicts (in which case it will take the first value)

mary-poppins commented:
The angular.io preview for cbe3748e60c1b1fa6b1af9a10e735c4ce0c7568a is available [here][1].

[1]: https://pr15621-cbe3748e60c1b1fa6b1af9a10e735c4ce0c7568a.ngbuilds.io/

mary-poppins commented:
The angular.io preview for d8eaf6f215074ab4f9641dfd724da1b1909eaa99 is available [here][1].

[1]: https://pr15621-d8eaf6f215074ab4f9641dfd724da1b1909eaa99.ngbuilds.io/

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -25,6 +25,21 @@ export function decimalDigest(message: i18n.Message): string {
 }
 
 /**
+ * Digest with a decimal function while ignoring placeholders and ICU expressions
```
We still depend on the placeholder name "ignoring placeholders" is not correct in that sense.

Could you please add more details ?

Please remove the comment about versions (version 1 here). Reason is digest and version are separate concern and this comment will become obsolete when the next version is out.

vicb commented on packages/compiler/src/i18n/digest.ts:
```diff
@@ -81,6 +96,16 @@ class _SerializerIgnoreIcuExpVisitor extends _SerializerVisitor {
+ * Serialize the i18n ast to something xml-like in order to generate an UID.
+ * Used by Id generation version 1
+ *
+ * Ignore the PH/ICU expressions so that message IDs stays identical if only the expression changes.
```
same comment as above for PH and version

vicb commented on packages/compiler/src/i18n/i18n_html_parser.ts:
```diff
@@ -55,20 +55,21 @@ export class I18NHtmlParser implements HtmlParser {
 
-function createSerializer(format?: string): Serializer {
+export function createSerializer(
+    format?: string, version: I18nVersion = I18nVersion.Version0): Serializer {
```
Why do we have a default parameter here ?
What about `export function createSerializer(format: string, version: I18nVersion)`
-> make both params mandatory

vicb commented on packages/compiler-cli/src/extractor.ts:
```diff
@@ -31,10 +33,9 @@ export class Extractor {
-    return promiseBundle.then(bundle => {
-      const content = this.serialize(bundle, formatName);
+    return this.extractBundle().then((bundle: compiler.MessageBundle) => {
+      const serializer = compiler.createSerializer(formatName);
```
pass an explicit version here.
(Probably we need to think about making this parameter - let's discuss this)

vicb commented on packages/compiler-cli/src/extractor.ts:
```diff
@@ -90,16 +75,16 @@ export class Extractor {
-      locale?: string|null, compilerHostContext?: CompilerHostContext,
-      ngCompilerHost?: CompilerHost): Extractor {
+      locale: string|null = null, compilerHostContext?: CompilerHostContext,
+      ngCompilerHost?: CompilerHost, version: I18nVersion = I18nVersion.Version0): Extractor {
```
changing the version here as no effect on extraction, right ?

_(the version is only passed to the `i18nHtmlParser` that does not use it for extraction)_

ocombe replied:
Yes, we always extract with the latest version, but this might be a mistake, shouldn't we let people use the old version for the extraction? in which case we should pass it here: https://github.com/ocombe/angular/blob/d8eaf6f215074ab4f9641dfd724da1b1909eaa99/packages/compiler-cli/src/extractor.ts#L37

vicb replied:
yep, this is my [comment above](https://github.com/angular/angular/pull/15621#pullrequestreview-42672948)

vicb commented on packages/compiler/src/i18n/i18n_html_parser.ts:
```diff
@@ -27,9 +27,9 @@ export class I18NHtmlParser implements HtmlParser {
       private _htmlParser: HtmlParser, translations?: string, translationsFormat?: string,
       missingTranslation: MissingTranslationStrategy = MissingTranslationStrategy.Warning,
-      console?: Console) {
+      console?: Console, version: I18nVersion = I18nVersion.Version0) {
```
remove default value here `= I18nVersion.Version0` ?

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,163 @@
+  const msgIdToElement = getElementsMapping(translations);
+  msgIdToElement.reverse().forEach(([id, element]: [string, ml.Element]) => {
+    const newId = mapping[id];
+    if (typeof newId !== 'undefined' && newId !== id) {
```
```
if (mapping.hasOwnProperty(id)) {
   const newId = mapping[id];
   // ...  
}
```

not sure if the test  `newId !== id` is really helpful here (There is no reason why we should produce this anyway, right ?)

ocombe replied:
true

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Updates translations based on the provided mapping
+ * @param translations
```
rename `translations` or add a comment that it might also be message.

Do we have support to migrate an xmb file ?

ocombe replied:
yes we can migrate all formats (including xmb)

vicb commented on packages/compiler-cli/src/codegen.ts:
```diff
@@ -70,37 +71,95 @@ export class CodeGenerator {
+/**
+ * @internal
+ */
+export function getTranslations(i18nFile: string | null): string {
```
rename to `readFileOrEmptyString`

ocombe replied:
k

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,163 @@
+  return {newToOld, oldToNew};
+}
+
+export function getElementsMapping(
```
what does this return ?
return an object ?
document ?

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,163 @@
+
+  const res: [string, ml.Element, string][] = [];
+  msgIdToElement.forEach(([id, element]: [string, ml.Element]) => {
+    res.push([id, element, msgMap[id] ? msgMap[id] : '']);
```
`.hasOwnProperty()`

vicb commented on packages/compiler/src/i18n/migrator.ts:
```diff
@@ -0,0 +1,163 @@
+    }
+  });
+
+  return {newToOld, oldToNew};
```
where do we need `oldToNew`

I think the names should be `newIdToOldIds` and `oldIdToNewId` (id vs ids is what is important here)

ocombe replied:
ok

vicb commented on packages/compiler-cli/src/migrator.ts:
```diff
@@ -0,0 +1,136 @@
+  private map(
+      mapping: {newToOld: {[newId: string]: string[]}, oldToNew: {[oldId: string]: string}},
+      i18nFile: string): string {
+    let dstPath = `${i18nFile.split('.').slice(0, -1).join('.')}_mapping.json`;
```
use `path` instead ?

ocombe replied:
we still need the name of the file to have a different mapping file for each file

vicb replied:
`path` the package to manipulate paths

vicb commented on packages/compiler-cli/src/migrator.ts:
```diff
@@ -0,0 +1,136 @@
+      mapping: {newToOld: {[newId: string]: string[]}, oldToNew: {[oldId: string]: string}},
+      i18nFile: string, formatName: string, version: I18nVersion,
+      resolve: string): Promise<string> {
+    const translations = getTranslations(i18nFile);
```
- I would separate I/O (reading files, prompt) from processing.
- It would still be nice to know conflicts even in the `resolve` mode (To display `no conflict detected / N conflicts have been solved automatically`)
- more comments would be nice

vicb replied:
Also probably the code related conflicts resolution should be in `@angular/compiler`
`compiler-cli` should only be orchestrating the calls

ocombe replied:
> I would separate I/O (reading files, prompt) from processing.
can be done

> It would still be nice to know conflicts even in the resolve mode (To display no conflict detected / N conflicts have been solved automatically)
I agree

> more comments would be nice
everytime I add comments you make me remove them :D I'll add some

> Also probably the code related conflicts resolution should be in @angular/compiler
compiler-cli should only be orchestrating the calls & dealing with I/O
it would complicate the code I think, the conflict resolution seems to be related to the prompts (for me) which are in the compiler-cli package

vicb commented on packages/compiler/src/config.ts:
```diff
@@ -21,7 +21,7 @@ export class CompilerConfig {
 
   constructor(
-      {defaultEncapsulation = ViewEncapsulation.Emulated, useJit = true, missingTranslation,
+      {defaultEncapsulation = ViewEncapsulation.Emulated, useJit = true, missingTranslation = null,
```
seems like this does not match the type ?

ocombe replied:
true, weird that the compiler didn't throw an error here

ocombe commented:
Closed in favor of https://github.com/angular/angular/pull/17638
